zulip/zulip-terminal
# Add DM panel to the left side view (rebased #1416)

### What does this PR do, and why?

This is an update on the current work in #1416.

Currently this only includes a few small fixups over #1416, that may be included or not:
- Make the new stream messages button not selectable (cannot receive focus)
- Adds a simple `>` prefix for the new Stream messages button as `STREAM_MESSAGE_MARKER`
- Make the DM messages button not selectable (cannot receive focus) [for now]

The way this is currently structured, this could be merged in two chunks (note that the summary of each chunk is not on a commit-wise basis):
- up to and including my `symbols/buttons` commit
  - adds the data for the total stream messages into a new (non-selectable) button
  - changes the left panel structure to have a simple dividing line from the top-left view buttons, rather than a `---Streams---` or `---Topics---` title
  - clarifies the context after that change by specifying 'Search topics' and 'Search streams' rather than the previous 'Search' in the panel search box
- that plus the rest of the commits
  - Moves the Direct messages button into a section with its own solid line separating it
  - Adds lots of model support for DM data, and updating it
  - Adjusts hotkeys so that from the left panel, `S` shows the view after the first PR chunk (above), with `P` switching to a view with the `P` view fully expanded
  - The fully expanded `P` view shows conversations with unreads, and user status
    (group conversations are included, unlike in the user list)
  - position in the DM list (and stream list) is preserved on switching between them

An additional UI 'chunk' could be to do the first bullet of the last chunk distinct from the data and view changes, but this change is currently combined with one of the later commits in the second chunk. That last chunk could also be split into data and UI updates, potentially.

### Feedback sought
Manual testing and feedback on:
1. the first chunk (eg. `git rebase -i` and 'edit' on the 'symbols/buttons' commit)
  - This is the anticipated UI direction to enable the full PR to merge
  - how does it feel during use?
  - does stream/topic switching look good?
  - how does it compare to `main` for you?
2. the full PR
  - This is the current plan for the first iteration of showing DM conversations
  - See previous conversations for motivation for this design:
    - https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/Private.20messages.20sidepanel.20.23T1416
    - https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/DM.20Side.20Panel
  - Note the outstanding (known) aspects below!
  - How does it look and work for you? Is anything clearly broken? any other improvements?

### Outstanding aspect(s)    


- [ ] `P` for DM feed functionality is currently removed, even from message context; it would be useful to retain this feature generally? [discuss]
  - [ ] Even the `P` DM feed button is not usable right now, for simplicity
- [ ] From the user list, `P` changes the left panel view, but doesn't re-focus; that's clearly buggy, but should it do something like that? [discuss]
- [ ] `S` now works differently in left panel (new view-change) vs message context (zoom-in); is that OK?
- [ ] While position in the stream list, and in any connected topic list, is preserved upon switching to/from the DM list, if the topic list was shown previously then it instead shows the stream list, which seems inconsistent? (eg. streams -`t`-> topics -`P`-> DMs -`S`-> streams)
- [ ] Sort user names in DM list like web app?
- [ ] Ensure DM list status symbols are updated
- [ ] Use 'huddle' symbols for conversations with >2 people (eg. ∷)

Note that not all of these are necessary for a full merge and could belong in follow-ups, but some relate to what seems like buggy, inconsistent or reduced behavior.

### External discussion & connections

- [x] Discussed in **#zulip-terminal** in `Private messages sidepanel #T1416` and `DM Side Panel`
- [ ] Fully fixes #
- [x] Partially fixes issue #1197 
- [x] Builds upon previous unmerged work in PR #1416 
- [ ] Is a follow-up to work in PR #
- [ ] Requires merge of PR #
- [ ] Merge will enable work on #

### How did you test this?

- [ ] Manually - Behavioral changes
- [ ] Manually - Visual changes
- [ ] Adapting existing automated tests
- [ ] Adding automated tests for new behavior (or missing tests)
- [ ] Existing automated tests should already cover this (*only a refactor of tested code*)

### Self-review checklist for each commit
- [ ] It is a [minimal coherent idea](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development)
- [ ] It has a commit summary following the [documented style](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development) (title & body)
- [ ] It has a commit summary describing the  motivation and reasoning for the change
- [ ] It individually passes linting and tests
- [ ] It contains test additions for any new behavior
- [ ] It flows clearly from a previous branch commit, and/or prepares for the next commit

### Visual changes
